Does not compile un MCH on mac (see below). Errors are unrelated to this commit.
```
2023-12-14@11:07:02:DEBUG:O2:O2:aliceo2: /Users/markusfasel/alice/sw/SOURCES/O2/aliceo2/0/Detectors/MUON/MCH/GlobalMapping/src/testChannelCode.cxx:40:39: error: implicit instantiation of undefined template 'std::array<o2::mch::ChannelCode, 2>'
2023-12-14@11:07:02:DEBUG:O2:O2:aliceo2:   std::array<o2::mch::ChannelCode, 2> ids = {
2023-12-14@11:07:02:DEBUG:O2:O2:aliceo2:                                       ^
2023-12-14@11:07:02:DEBUG:O2:O2:aliceo2: /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.1.sdk/usr/include/c++/v1/__tuple:219:64: note: template is declared here
2023-12-14@11:07:02:DEBUG:O2:O2:aliceo2: template <class _Tp, size_t _Size> struct _LIBCPP_TEMPLATE_VIS array;
2023-12-14@11:07:02:DEBUG:O2:O2:aliceo2:                                                                ^
2023-12-14@11:07:02:DEBUG:O2:O2:aliceo2: /Users/markusfasel/alice/sw/SOURCES/O2/aliceo2/0/Detectors/MUON/MCH/GlobalMapping/src/testChannelCode.cxx:58:39: error: implicit instantiation of undefined template 'std::array<o2::mch::ChannelCode, 2>'
2023-12-14@11:07:02:DEBUG:O2:O2:aliceo2:   std::array<o2::mch::ChannelCode, 2> ids = {
2023-12-14@11:07:02:DEBUG:O2:O2:aliceo2:                                       ^
2023-12-14@11:07:02:DEBUG:O2:O2:aliceo2: /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.1.sdk/usr/include/c++/v1/__tuple:219:64: note: template is declared here
2023-12-14@11:07:02:DEBUG:O2:O2:aliceo2: template <class _Tp, size_t _Size> struct _LIBCPP_TEMPLATE_VIS array;
2023-12-14@11:07:02:DEBUG:O2:O2:aliceo2:                                                                ^
2023-12-14@11:07:02:DEBUG:O2:O2:aliceo2: 2 errors generated.
```

--- a/DataFormats/Detectors/CTP/include/DataFormatsCTP/Configuration.h
+++ b/DataFormats/Detectors/CTP/include/DataFormatsCTP/Configuration.h
@@ -118,6 +118,7 @@ struct CTPClass {
 struct CTPInputsConfiguration {
   CTPInputsConfiguration() = default;
   std::vector<CTPInput> CTPInputs;
+  static const std::vector<CTPInput> CTPInputsDefault;
   int createInputsConfigFromFile(std::string& filename);
   void printStream(std::ostream& strem) const;
   static CTPInputsConfiguration defaultInputConfig;

--- a/DataFormats/Detectors/CTP/src/Configuration.cxx
+++ b/DataFormats/Detectors/CTP/src/Configuration.cxx
@@ -1044,6 +1044,18 @@ void CTPInputsConfiguration::printStream(std::ostream& stream) const
     input.printStream(stream);
   }
 }
+//
+// EMBA - software generated input for EMC - Min Bias Accepted
+//
+const std::vector<CTPInput> CTPInputsConfiguration::CTPInputsDefault =
+  {
+    CTPInput("MT0A", "FT0", 1), CTPInput("MT0C", "FT0", 2), CTPInput("MTVX", "FT0", 3), CTPInput("MTSC", "FT0", 4), CTPInput("MTCE", "FT0", 5),
+    CTPInput("MVBA", "FV0", 6), CTPInput("MVOR", "FV0", 7), CTPInput("MVIR", "FV0", 8), CTPInput("MVNC", "FV0", 9), CTPInput("MVCH", "FV0", 10),
+    CTPInput("0UCE", "FDD", 13), CTPInput("0USC", "FDD", 15), CTPInput("0UVX", "FDD", 16), CTPInput("0U0C", "FDD", 17), CTPInput("0U0A", "FDD", 18),
+    CTPInput("0DMC", "EMC", 14), CTPInput("0DJ1", "EMC", 41), CTPInput("0DG1", "EMC", 42), CTPInput("0DJ2", "EMC", 43), CTPInput("0DG2", "EMC", 44),
+    CTPInput("0EMC", "EMC", 21), CTPInput("0EJ1", "EMC", 37), CTPInput("0EG1", "EMC", 38), CTPInput("0EJ2", "EMC", 39), CTPInput("0EG2", "EMC", 40),
+    CTPInput("0PH0", "PHS", 22), CTPInput("1PHL", "PHS", 27), CTPInput("1PHH", "PHS", 28), CTPInput("1PHL", "PHM", 29),
+    CTPInput("1ZED", "ZDC", 25), CTPInput("1ZNC", "ZDC", 26), CTPInput("EMBA", "EMC", 48)};
 void CTPInputsConfiguration::initDefaultInputConfig()
 {
   defaultInputConfig.CTPInputs.push_back(CTPInput("MT0A", "FT0", 1));

--- a/Detectors/CTP/simulation/src/Digitizer.cxx
+++ b/Detectors/CTP/simulation/src/Digitizer.cxx
@@ -27,7 +27,9 @@ std::vector<CTPDigit> Digitizer::process(const gsl::span<o2::ctp::CTPInputDigit>
   std::map<o2::detectors::DetID::ID, std::vector<CTPInput>> det2ctpinp = mCTPConfiguration->getDet2InputMap();
   // To be taken from config database ?
   std::map<std::string, uint64_t> detInputName2Mask =
-    {{"MVBA", 1}, {"MVIR", 0x10}, {"MVOR", 2}, {"MVNC", 4}, {"MVCH", 8}, {"MT0A", 1}, {"MT0C", 2}, {"MTVX", 0x10}, {"MTCE", 8}, {"MTSC", 0x4}, {"0U0A", 1}, {"0U0C", 2}, {"0UVX", 0x10}, {"0UCE", 8}, {"0USC", 0x4}};
+    {{"MVBA", 1}, {"MVOR", 2}, {"MVNC", 4}, {"MVCH", 8}, {"MVIR", 0x10}, {"MT0A", 1}, {"MT0C", 2}, {"MTSC", 4}, {"MTCE", 8}, {"MTVX", 0x10}, {"0U0A", 1}, {"0U0C", 2}, {"0USC", 4}, {"0UCE", 8}, {"0UVX", 0x10}, {"EMBA", 0x1}, {"0EMC", 0x2}, {"0DMC", 0x4}};
+
+  // pre-sorting detector inputs per interaction record
   std::map<o2::InteractionRecord, std::vector<const CTPInputDigit*>> predigits;
   for (auto const& inp : detinputs) {
     predigits[inp.intRecord].push_back(&inp);
@@ -62,10 +64,23 @@ std::vector<CTPDigit> Digitizer::process(const gsl::span<o2::ctp::CTPInputDigit>
           break;
         }
         case o2::detectors::DetID::EMC: {
-          for (auto const& ctpinp : det2ctpinp[o2::detectors::DetID::EMC]) {
-            uint64_t mask = (inp->inputsMask).to_ullong() & detInputName2Mask[ctpinp.name];
-            inpmaskcoll |= std::bitset<CTP_NINPUTS>(ctpinp.inputMask);
+          // uint64_t inpmaskdebug = 1;
+          uint64_t inpmaskdebug = (inp->inputsMask).to_ullong();
+          if (inpmaskdebug & detInputName2Mask["EMBA"]) {
+            // MB-accept must be treated separately, as it is not a CTP input
+            std::bitset<CTP_NINPUTS> emcMBaccept;
+            emcMBaccept.set(CTP_NINPUTS - 1, 1);
+            inpmaskcoll |= emcMBaccept;
+          } else {
+            for (auto const& ctpinp : det2ctpinp[o2::detectors::DetID::EMC]) {
+              uint64_t mask = inpmaskdebug & detInputName2Mask[ctpinp.name];
+              // uint64_t mask = (inp->inputsMask).to_ullong() & detInputName2Mask[ctpinp.name];
+              if (mask) {
+                inpmaskcoll |= std::bitset<CTP_NINPUTS>(ctpinp.inputMask);
+              }
+            }
           }
+          LOG(info) << "EMC input mask:" << inpmaskcoll;
           break;
         }
         case o2::detectors::DetID::PHS: {
@@ -87,11 +102,24 @@ std::vector<CTPDigit> Digitizer::process(const gsl::span<o2::ctp::CTPInputDigit>
           LOG(error) << "CTP Digitizer: unknown detector:" << inp->detector;
           break;
       }
+      // inpmaskcoll.reset();  // debug
+      // inpmaskcoll[47] = 1;  // debug
+    } // end loop over trigger input for this interaction
+    if (inpmaskcoll.to_ullong()) {
+      // we put the trigger only when non-trivial
+      std::bitset<64> classmask;
+      calculateClassMask(inpmaskcoll, classmask);
+      if (classmask.to_ulong() == 0) {
+        // No class accepted
+        continue;
+      }
+      CTPDigit data;
+      data.intRecord = hits.first;
+      data.CTPInputMask = inpmaskcoll;
+      data.CTPClassMask = classmask;
+      digits.emplace_back(data);
+      LOG(info) << "Trigger-Event " << data.intRecord.bc << " " << data.intRecord.orbit << " Input mask:" << inpmaskcoll;
     }
-    LOG(info) << data.intRecord.bc << " " << data.intRecord.orbit << " Input mask:" << inpmaskcoll;
-    data.CTPInputMask = inpmaskcoll;
-    calculateClassMask(inpmaskcoll, data.CTPClassMask);
-    digits.emplace_back(data);
   }
   return std::move(digits);
 }
@@ -99,10 +127,39 @@ void Digitizer::calculateClassMask(const std::bitset<CTP_NINPUTS> ctpinpmask, st
 {
   classmask = 0;
   for (auto const& tcl : mCTPConfiguration->getCTPClasses()) {
-    if (tcl.descriptor->getInputsMask() & ctpinpmask.to_ullong()) {
-      classmask |= (1 << tcl.classMask);
+    if (tcl.cluster->name == "emc") {
+      // check if Min Bias EMC class
+      bool tvxMBemc = tcl.name.find("C0TVX-B-NOPF-EMC") != std::string::npos; // 2023
+      tvxMBemc |= tcl.name.find("C0TVX-A-NOPF-EMC") != std::string::npos;
+      tvxMBemc |= tcl.name.find("C0TVX-C-NOPF-EMC") != std::string::npos;
+      tvxMBemc |= tcl.name.find("C0TVX-E-NOPF-EMC") != std::string::npos;
+      if (tcl.cluster->name == "emc") {
+        tvxMBemc |= tcl.name.find("minbias_TVX_L0") != std::string::npos; // 2022
+      }
+      if (tcl.descriptor->getInputsMask() & ctpinpmask.to_ullong()) {
+        // require real physics input in any case
+        if (tvxMBemc) {
+          // if the class is a min. bias class accept it only if the MB-accept bit is set in addition
+          // (fake trigger input)
+          if (ctpinpmask[CTP_NINPUTS - 1]) {
+            classmask |= tcl.classMask;
+            LOG(info) << "adding MBA:" << tcl.name;
+          }
+        } else {
+          // EMCAL rare triggers - physical trigger input
+          // class identification can be handled like in the case of the other
+          // classes as EMCAL trigger input is required
+          classmask |= tcl.classMask;
+        }
+      }
+    } else {
+      if (tcl.descriptor->getInputsMask() & ctpinpmask.to_ullong()) {
+        classmask |= tcl.classMask;
+      }
     }
   }
+  LOG(info) << "input mask:" << ctpinpmask;
+  LOG(info) << "class mask:" << classmask;
 }
 void Digitizer::init()
 {

--- a/Detectors/EMCAL/simulation/include/EMCALSimulation/Digitizer.h
+++ b/Detectors/EMCAL/simulation/include/EMCALSimulation/Digitizer.h
@@ -94,21 +94,28 @@ class Digitizer : public TObject
   const std::vector<o2::emcal::TriggerRecord>& getTriggerRecords() const { return mDigits.getTriggerRecords(); }
   const o2::dataformats::MCTruthContainer<o2::emcal::MCLabel>& getMCLabels() const { return mDigits.getMCLabels(); }
 
+  static constexpr int getTOFSamplingBins() { return EMC_TOF_BINS; }
+
  private:
-  static constexpr int EMC_PHASES = 4; ///< Number of phases
-  short mEventTimeOffset = 0;          ///< event time difference from trigger time (in number of bins)
-  short mPhase = 0;                    ///< event phase
-  UInt_t mROFrameMin = 0;              ///< lowest RO frame of current digits
-  UInt_t mROFrameMax = 0;              ///< highest RO frame of current digits
-  bool mSmearEnergy = true;            ///< do time and energy smearing
-  bool mSimulateTimeResponse = true;   ///< simulate time response
-  const SimParam* mSimParam = nullptr; ///< SimParam object
+  using TimeSampleContainer = std::array<double, constants::EMCAL_MAXTIMEBINS>;
+  static constexpr int EMC_PHASES = 4;                                                  ///< Number of phases
+  static constexpr int EMC_TOF_BINS = 1500;                                             ///< Number of bins in TOF sampling of the time response
+  static constexpr double EMC_TOF_MIN = 0;                                              ///< Min TOF
+  static constexpr double EMC_TOF_MAX = 1500.;                                          ///< Max TOF
+  static constexpr double EMC_TOF_BINWITH = (EMC_TOF_MAX - EMC_TOF_MIN) / EMC_TOF_BINS; ///< Number time samples simulated
+  short mEventTimeOffset = 0;                                                           ///< event time difference from trigger time (in number of bins)
+  short mPhase = 0;                                                                     ///< event phase
+  UInt_t mROFrameMin = 0;                                                               ///< lowest RO frame of current digits
+  UInt_t mROFrameMax = 0;                                                               ///< highest RO frame of current digits
+  bool mSmearEnergy = true;                                                             ///< do time and energy smearing
+  bool mSimulateTimeResponse = true;                                                    ///< simulate time response
+  const SimParam* mSimParam = nullptr;                                                  ///< SimParam object
 
   std::vector<Digit> mTempDigitVector;     ///< temporary digit storage
   o2::emcal::DigitsWriteoutBuffer mDigits; ///< used to sort digits and labels by tower
 
   TRandom3* mRandomGenerator = nullptr; ///< random number generator
-  std::array<std::array<double, constants::EMCAL_MAXTIMEBINS>, EMC_PHASES>
+  std::array<std::array<TimeSampleContainer, EMC_TOF_BINS>, EMC_PHASES>
     mAmplitudeInTimeBins; ///< template of the sampled time response function: amplitude of signal for each time bin (per phase)
 
   std::unique_ptr<o2::utils::TreeStreamRedirector> mDebugStream = nullptr;

--- a/Detectors/EMCAL/simulation/include/EMCALSimulation/Digitizer.h
+++ b/Detectors/EMCAL/simulation/include/EMCALSimulation/Digitizer.h
@@ -67,7 +67,7 @@ class Digitizer : public TObject
   /// Steer conversion of hits to digits
   void process(const std::vector<LabeledDigit>& labeledDigit);
 
-  void setEventTime(o2::InteractionTimeRecord record);
+  void setEventTime(o2::InteractionTimeRecord record, bool trigger);
   double getTriggerTime() const { return mDigits.getTriggerTime(); }
   double getEventTime() const { return mDigits.getEventTime(); }
   bool isLive(double t) const { return mDigits.isLive(t); }

--- a/Detectors/EMCAL/simulation/include/EMCALSimulation/DigitsWriteoutBuffer.h
+++ b/Detectors/EMCAL/simulation/include/EMCALSimulation/DigitsWriteoutBuffer.h
@@ -81,7 +81,7 @@ class DigitsWriteoutBuffer
   unsigned int getBufferSize() const { return mBufferSize; }
 
   /// forward the marker for every 100 ns
-  void forwardMarker(o2::InteractionTimeRecord record);
+  void forwardMarker(o2::InteractionTimeRecord record, bool trigger);
 
   /// Setters for the live time, busy time, pre-trigger time
   void setLiveTime(unsigned int liveTime) { mLiveTime = liveTime; }

--- a/Detectors/EMCAL/simulation/src/Digitizer.cxx
+++ b/Detectors/EMCAL/simulation/src/Digitizer.cxx
@@ -64,9 +64,12 @@ void Digitizer::init()
       // parameter 1: Handling phase + delay
       // phase: 25 ns * phase index (-4)
       // delay: Average signal delay
-      RawResponse.SetParameter(1, 0.25 * phase + mSimParam->getSignalDelay() / constants::EMCAL_TIMESAMPLE);
-      for (int sample = 0; sample < constants::EMCAL_MAXTIMEBINS; sample++) {
-        mAmplitudeInTimeBins[phase][sample] = RawResponse.Eval(sample);
+      for (int itofbin = 0; itofbin < EMC_TOF_BINS; itofbin++) {
+        double tofbincenter = itofbin * EMC_TOF_BINWITH + 0.5 * EMC_TOF_BINWITH;
+        RawResponse.SetParameter(1, 0.25 * phase + (tofbincenter + mSimParam->getSignalDelay()) / constants::EMCAL_TIMESAMPLE);
+        for (int sample = 0; sample < constants::EMCAL_MAXTIMEBINS; sample++) {
+          mAmplitudeInTimeBins[phase][itofbin][sample] = RawResponse.Eval(sample);
+        }
       }
     }
   } else {
@@ -158,15 +161,29 @@ void Digitizer::sampleSDigit(const Digit& sDigit)
 
   Double_t energies[15];
   if (mSimulateTimeResponse) {
-    for (int sample = 0; sample < mAmplitudeInTimeBins[mPhase].size(); sample++) {
+    if (sDigit.getTimeStamp() + mSimParam->getSignalDelay() + mPhase * 25 > EMC_TOF_MAX) {
+      // Digit time larger than sampling window, will not be sampled
+      // For time response simulation take also signal delay and phase into account
+      return;
+    }
+    int tofbin = static_cast<int>(sDigit.getTimeStamp() / EMC_TOF_BINWITH);
+    if (tofbin >= EMC_TOF_BINS) {
+      tofbin = EMC_TOF_BINS - 1;
+    }
+    for (int sample = 0; sample < mAmplitudeInTimeBins[mPhase][tofbin].size(); sample++) {
 
-      double val = energy * (mAmplitudeInTimeBins[mPhase][sample]);
+      double val = energy * (mAmplitudeInTimeBins[mPhase][tofbin][sample]);
       energies[sample] = val;
       double digitTime = mEventTimeOffset * constants::EMCAL_TIMESAMPLE;
       Digit digit(tower, val, digitTime);
       mTempDigitVector.push_back(digit);
     }
   } else {
+    if (sDigit.getTimeStamp() > EMC_TOF_MAX) {
+      // Digit time larger than sampling window, will not be sampled
+      // In non-sampled mode only apply the max. time window
+      return;
+    }
     Digit digit(tower, energy, smearTime(sDigit.getTimeStamp(), energy));
     mTempDigitVector.push_back(digit);
   }

--- a/Detectors/EMCAL/simulation/src/Digitizer.cxx
+++ b/Detectors/EMCAL/simulation/src/Digitizer.cxx
@@ -228,10 +228,10 @@ double Digitizer::smearTime(double time, double energy)
 }
 
 //_______________________________________________________________________
-void Digitizer::setEventTime(o2::InteractionTimeRecord record)
+void Digitizer::setEventTime(o2::InteractionTimeRecord record, bool trigger)
 {
 
-  mDigits.forwardMarker(record);
+  mDigits.forwardMarker(record, trigger);
 
   mPhase = mSimParam->doSimulateL1Phase() ? mDigits.getPhase() : 0;
 

--- a/Detectors/EMCAL/simulation/src/DigitsVectorStream.cxx
+++ b/Detectors/EMCAL/simulation/src/DigitsVectorStream.cxx
@@ -121,7 +121,7 @@ void DigitsVectorStream::fill(std::deque<o2::emcal::DigitTimebin>& digitlist, o2
   mTriggerRecords.emplace_back(record, o2::trigger::PhT, mStartIndex, numberOfNewDigits);
   mStartIndex = mDigits.size();
 
-  LOG(info) << "Have " << mStartIndex << " digits ";
+  LOG(info) << "Trigger Orbit " << record.orbit << ", BC " << record.bc << ": have " << numberOfNewDigits << " digits (" << mStartIndex << " total)";
 }
 
 //_______________________________________________________________________

--- a/Detectors/EMCAL/simulation/src/DigitsWriteoutBuffer.cxx
+++ b/Detectors/EMCAL/simulation/src/DigitsWriteoutBuffer.cxx
@@ -19,6 +19,7 @@
 #include "EMCALSimulation/DigitsWriteoutBuffer.h"
 #include "CommonDataFormat/InteractionRecord.h"
 #include "TMath.h"
+#include <fairlogger/Logger.h>
 
 using namespace o2::emcal;
 
@@ -80,7 +81,7 @@ void DigitsWriteoutBuffer::addDigits(unsigned int towerID, std::vector<LabeledDi
 // is removed and pushed at the end of the past dequeue. At the same time a new entry in
 // the future dequeue is pushed at the end, and - in case the past dequeue reached 15 entries
 // the first entry of the past dequeue is removed.
-void DigitsWriteoutBuffer::forwardMarker(o2::InteractionTimeRecord record)
+void DigitsWriteoutBuffer::forwardMarker(o2::InteractionTimeRecord record, bool trigger)
 {
 
   unsigned long eventTime = record.getTimeNS();
@@ -124,7 +125,8 @@ void DigitsWriteoutBuffer::forwardMarker(o2::InteractionTimeRecord record)
 
   // If we have a trigger, all the time bins in the future buffer will be set to record mode
   // the last time bin will the end of the readout window since it will be mTriggerTime + 1500 ns
-  if ((eventTime - mTriggerTime) >= (mLiveTime + mBusyTime) || mFirstEvent) {
+  if (trigger && ((eventTime - mTriggerTime) >= (mLiveTime + mBusyTime) || mFirstEvent)) {
+    LOG(debug) << "Trigger received and accepted";
     mTriggerTime = eventTime;
     mTimedDigitsFuture.front().mTriggerColl = true;
     mTimedDigitsFuture.front().mInterRecord = record;
@@ -142,8 +144,8 @@ void DigitsWriteoutBuffer::forwardMarker(o2::InteractionTimeRecord record)
   }
 
   // If we have a pre-trigger collision, all the time bins in the future buffer will be set to record mode
-  if ((eventTime - mTriggerTime) >= (mLiveTime + mBusyTime - mPreTriggerTime)) {
-    std::cout << "Pre-trigger collision\n";
+  if ((eventTime - mTriggerTime) >= (mLiveTime + mBusyTime - mPreTriggerTime) || mFirstEvent) {
+    LOG(debug) << "Pre-trigger collision";
     long timeStamp = (eventTime / 100) * 100; /// This is to make the event time multiple of 100s
     for (auto& iNode : mTimedDigitsFuture) {
       long diff = (timeStamp - eventTime);

--- a/Detectors/EMCAL/workflow/CMakeLists.txt
+++ b/Detectors/EMCAL/workflow/CMakeLists.txt
@@ -25,8 +25,20 @@ o2_add_library(EMCALWorkflow
         src/CellRecalibratorSpec.cxx
         src/EntropyDecoderSpec.cxx
         src/StandaloneAODProducerSpec.cxx
-        PUBLIC_LINK_LIBRARIES O2::Framework O2::DataFormatsCTP O2::DataFormatsEMCAL O2::EMCALSimulation O2::Steer
-        O2::DPLUtils O2::EMCALBase O2::EMCALCalib O2::EMCALReconstruction O2::Algorithm O2::MathUtils)
+        PUBLIC_LINK_LIBRARIES O2::Framework
+        O2::DataFormatsCTP
+        O2::DataFormatsEMCAL
+        O2::DataFormatsFT0
+        O2::DataFormatsFV0
+        O2::EMCALSimulation
+        O2::Steer
+        O2::DPLUtils
+        O2::EMCALBase
+        O2::EMCALCalib
+        O2::EMCALCalibration
+        O2::EMCALReconstruction
+        O2::Algorithm
+        O2::MathUtils)
 
 o2_add_executable(reco-workflow
         COMPONENT_NAME emcal

--- a/Detectors/EMCAL/workflow/include/EMCALWorkflow/EMCALDigitizerSpec.h
+++ b/Detectors/EMCAL/workflow/include/EMCALWorkflow/EMCALDigitizerSpec.h
@@ -70,6 +70,7 @@ class DigitizerSpec final : public o2::base::BaseDPLDigitizer, public o2::framew
  private:
   Bool_t mFinished = false;                   ///< Flag for digitization finished
   bool mIsConfigured = false;                 ///< Initialization status of the digitizer
+  bool mRunSDitizer = false;                  ///< Run SDigitization
   Digitizer mDigitizer;                       ///< Digitizer object
   o2::emcal::SDigitizer mSumDigitizer;        ///< Summed digitizer
   std::shared_ptr<CalibLoader> mCalibHandler; ///< Handler of calibration objects

--- a/Detectors/EMCAL/workflow/include/EMCALWorkflow/EMCALDigitizerSpec.h
+++ b/Detectors/EMCAL/workflow/include/EMCALWorkflow/EMCALDigitizerSpec.h
@@ -28,6 +28,12 @@ class TChain;
 
 namespace o2
 {
+
+namespace ctp
+{
+class CTPConfiguration;
+}
+
 namespace emcal
 {
 class CalibLoader;
@@ -45,7 +51,7 @@ class DigitizerSpec final : public o2::base::BaseDPLDigitizer, public o2::framew
  public:
   using o2::base::BaseDPLDigitizer::init;
   /// \brief Constructor
-  DigitizerSpec(std::shared_ptr<CalibLoader> calibloader) : o2::base::BaseDPLDigitizer(o2::base::InitServices::GEOM), o2::framework::Task(), mCalibHandler(calibloader) {}
+  DigitizerSpec(std::shared_ptr<CalibLoader> calibloader, bool requireCTPInput) : o2::base::BaseDPLDigitizer(o2::base::InitServices::GEOM), o2::framework::Task(), mRequireCTPInput(requireCTPInput), mCalibHandler(calibloader) {}
 
   /// \brief Destructor
   ~DigitizerSpec() final = default;
@@ -71,16 +77,18 @@ class DigitizerSpec final : public o2::base::BaseDPLDigitizer, public o2::framew
   Bool_t mFinished = false;                   ///< Flag for digitization finished
   bool mIsConfigured = false;                 ///< Initialization status of the digitizer
   bool mRunSDitizer = false;                  ///< Run SDigitization
+  bool mRequireCTPInput = false;              ///< Require CTP min. bias input
   Digitizer mDigitizer;                       ///< Digitizer object
   o2::emcal::SDigitizer mSumDigitizer;        ///< Summed digitizer
   std::shared_ptr<CalibLoader> mCalibHandler; ///< Handler of calibration objects
   std::vector<Hit> mHits;                     ///< Vector with input hits
   std::vector<TChain*> mSimChains;
+  o2::ctp::CTPConfiguration* mCTPConfig; ///< CTP configuration
 };
 
 /// \brief Create new digitizer spec
 /// \return Digitizer spec
-o2::framework::DataProcessorSpec getEMCALDigitizerSpec(int channel, bool mctruth = true, bool useccdb = true);
+o2::framework::DataProcessorSpec getEMCALDigitizerSpec(int channel, bool requireCTPInput, bool mctruth = true, bool useccdb = true);
 
 } // namespace emcal
 } // end namespace o2

--- a/Detectors/EMCAL/workflow/src/EMCALDigitizerSpec.cxx
+++ b/Detectors/EMCAL/workflow/src/EMCALDigitizerSpec.cxx
@@ -119,7 +119,18 @@ void DigitizerSpec::run(framework::ProcessingContext& ctx)
 
       LOG(info) << "For collision " << collID << " eventID " << part.entryID << " found " << mHits.size() << " hits ";
 
-      std::vector<o2::emcal::LabeledDigit> summedDigits = mSumDigitizer.process(mHits);
+      std::vector<o2::emcal::LabeledDigit> summedDigits;
+      if (mRunSDitizer) {
+        summedDigits = mSumDigitizer.process(mHits);
+      } else {
+        for (auto& hit : mHits) {
+          o2::emcal::MCLabel digitlabel(hit.GetTrackID(), part.entryID, part.sourceID, false, 1.);
+          if (hit.GetEnergyLoss() < __DBL_EPSILON__) {
+            digitlabel.setAmplitudeFraction(0);
+          }
+          summedDigits.emplace_back(hit.GetDetectorID(), hit.GetEnergyLoss(), hit.GetTime(), digitlabel);
+        }
+      }
 
       // call actual digitization procedure
       mDigitizer.process(summedDigits);

--- a/Detectors/EMCAL/workflow/src/EMCALDigitizerSpec.cxx
+++ b/Detectors/EMCAL/workflow/src/EMCALDigitizerSpec.cxx
@@ -27,8 +27,11 @@
 #include "CommonDataFormat/EvIndex.h"
 #include "DetectorsCommonDataFormats/DetID.h"
 #include "DataFormatsParameters/GRPObject.h"
+#include "DataFormatsCTP/Configuration.h"
 #include "DataFormatsCTP/Digits.h"
 #include "DataFormatsEMCAL/TriggerRecord.h"
+#include "DataFormatsFT0/Digit.h"
+#include "DataFormatsFV0/Digit.h"
 
 using namespace o2::framework;
 using SubSpecificationType = o2::framework::DataAllocator::SubSpecificationType;
@@ -91,6 +94,77 @@ void DigitizerSpec::run(framework::ProcessingContext& ctx)
   TStopwatch timer;
   timer.Start();
 
+  // Load FIT triggers if not running in self-triggered mode
+  std::vector<o2::InteractionRecord> mbtriggers;
+  if (mRequireCTPInput) {
+    // Hopefully at some point we can replace it by CTP input digits
+    // In case of CTP digits react only to trigger inputs activated in trigger configuration
+    // For the moment react to FIT vertex, cent and semicent triggers
+    ctx.inputs().get<o2::ctp::CTPConfiguration*>("ctpconfig");
+    std::vector<uint64_t> inputmasks;
+    for (const auto& trg : mCTPConfig->getCTPClasses()) {
+      if (trg.cluster->maskCluster[o2::detectors::DetID::EMC]) {
+        // Class triggering EMCAL cluster
+        LOG(debug) << "Found trigger class for EMCAL cluster: " << trg.name << " with input mask " << std::bitset<64>(trg.descriptor->getInputsMask());
+        inputmasks.emplace_back(trg.descriptor->getInputsMask());
+      }
+    }
+    unsigned long ft0mask = 0, fv0mask = 0;
+    std::map<std::string, uint64_t> detInputName2Mask =
+      {{"MVBA", 1}, {"MVOR", 2}, {"MVNC", 4}, {"MVCH", 8}, {"MVIR", 0x10}, {"MT0A", 1}, {"MT0C", 2}, {"MTSC", 4}, {"MTCE", 8}, {"MTVX", 0x10}};
+    // Translation 2022: EMCAL cluster received CTP input masks, need to track it to FIT trigger masks
+    std::map<std::string, std::pair<o2::detectors::DetID, std::string>> ctpInput2DetInput = {
+      {"0VBA", {o2::detectors::DetID::FV0, "MVBA"}}, {"0VOR", {o2::detectors::DetID::FV0, "MVOR"}}, {"0VNC", {o2::detectors::DetID::FV0, "MVNC"}}, {"0VCH", {o2::detectors::DetID::FV0, "MVCH"}}, {"0VIR", {o2::detectors::DetID::FV0, "MVIR"}}, {"0T0A", {o2::detectors::DetID::FT0, "MT0A"}}, {"0T0C", {o2::detectors::DetID::FT0, "MT0C"}}, {"0TSC", {o2::detectors::DetID::FT0, "MTSC"}}, {"0TCE", {o2::detectors::DetID::FT0, "MTCE"}}, {"0TVX", {o2::detectors::DetID::FT0, "MTVX"}}};
+    for (const auto& [det, ctpinputs] : mCTPConfig->getDet2InputMap()) {
+      if (!(det == o2::detectors::DetID::FT0 || det == o2::detectors::DetID::FV0 || det == o2::detectors::DetID::CTP)) {
+        continue;
+      }
+      for (const auto& input : ctpinputs) {
+        LOG(debug) << "CTP det input: " << input.name << " with mask " << std::bitset<64>(input.inputMask);
+        bool isSelected = false;
+        for (auto testmask : inputmasks) {
+          if (testmask & input.inputMask) {
+            isSelected = true;
+          }
+        }
+        if (isSelected) {
+          std::string usedInputName = input.name;
+          o2::detectors::DetID usedDetID = det;
+          if (det == o2::detectors::DetID::CTP) {
+            auto found = ctpInput2DetInput.find(input.name);
+            if (found != ctpInput2DetInput.end()) {
+              usedInputName = found->second.second;
+              usedDetID = found->second.first;
+              LOG(debug) << "Decoded " << input.name << " -> " << usedInputName;
+            }
+          }
+          auto maskFound = detInputName2Mask.find(usedInputName);
+          if (maskFound != detInputName2Mask.end()) {
+            if (usedDetID == o2::detectors::DetID::FT0) {
+              ft0mask |= maskFound->second;
+            } else {
+              fv0mask |= maskFound->second;
+            }
+          }
+        }
+      }
+    }
+    LOG(debug) << "FTO mask: " << std::bitset<64>(ft0mask);
+    LOG(debug) << "FVO mask: " << std::bitset<64>(fv0mask);
+    for (const auto& trg : ctx.inputs().get<gsl::span<o2::ft0::DetTrigInput>>("ft0inputs")) {
+      if (trg.mInputs.to_ulong() & ft0mask) {
+        mbtriggers.emplace_back(trg.mIntRecord);
+      }
+    }
+    for (const auto& trg : ctx.inputs().get<gsl::span<o2::fv0::DetTrigInput>>("fv0inputs")) {
+      if (trg.mInputs.to_ulong() & fv0mask) {
+        if (std::find(mbtriggers.begin(), mbtriggers.end(), trg.mIntRecord) == mbtriggers.end()) {
+          mbtriggers.emplace_back(trg.mIntRecord);
+        }
+      }
+    }
+  }
+
   LOG(info) << " CALLING EMCAL DIGITIZATION ";
   o2::dataformats::MCTruthContainer<o2::emcal::MCLabel> labelAccum;
 
@@ -99,8 +173,13 @@ void DigitizerSpec::run(framework::ProcessingContext& ctx)
   // loop over all composite collisions given from context
   // (aka loop over all the interaction records)
   for (int collID = 0; collID < timesview.size(); ++collID) {
+    bool trigger = true; // Default: Self-triggered mode - all collisions treated as trigger
+    if (mRequireCTPInput) {
+      // check if we have a trigger input from CTP
+      trigger = (std::find(mbtriggers.begin(), mbtriggers.end(), timesview[collID]) != mbtriggers.end());
+    }
 
-    mDigitizer.setEventTime(timesview[collID]);
+    mDigitizer.setEventTime(timesview[collID], trigger);
 
     if (!mDigitizer.isLive()) {
       continue;
@@ -180,9 +259,13 @@ void DigitizerSpec::finaliseCCDB(o2::framework::ConcreteDataMatcher& matcher, vo
   if (mCalibHandler->finalizeCCDB(matcher, obj)) {
     return;
   }
+  if (matcher == o2::framework::ConcreteDataMatcher("CTP", "CTPCONFIG", 0)) {
+    std::cout << "Loading CTP configuration" << std::endl;
+    mCTPConfig = reinterpret_cast<o2::ctp::CTPConfiguration*>(obj);
+  }
 }
 
-o2::framework::DataProcessorSpec getEMCALDigitizerSpec(int channel, bool mctruth, bool useccdb)
+o2::framework::DataProcessorSpec getEMCALDigitizerSpec(int channel, bool requireCTPInput, bool mctruth, bool useccdb)
 {
   // create the full data processor spec using
   //  a name identifier
@@ -206,12 +289,17 @@ o2::framework::DataProcessorSpec getEMCALDigitizerSpec(int channel, bool mctruth
     calibloader->enableSimParams(true);
     calibloader->defineInputSpecs(inputs);
   }
+  if (requireCTPInput) {
+    inputs.emplace_back("ft0inputs", "FT0", "TRIGGERINPUT", 0, Lifetime::Timeframe);
+    inputs.emplace_back("fv0inputs", "FV0", "TRIGGERINPUT", 0, Lifetime::Timeframe);
+    inputs.emplace_back("ctpconfig", "CTP", "CTPCONFIG", 0, Lifetime::Condition, ccdbParamSpec("CTP/Config/Config", true));
+  }
 
   return DataProcessorSpec{
     "EMCALDigitizer", // Inputs{InputSpec{"collisioncontext", "SIM", "COLLISIONCONTEXT", static_cast<SubSpecificationType>(channel), Lifetime::Timeframe}, InputSpec{"EMC_SimParam", o2::header::gDataOriginEMC, "SIMPARAM", 0, Lifetime::Condition, ccdbParamSpec("EMC/Config/SimParam")}},
     inputs,
     outputs,
-    AlgorithmSpec{o2::framework::adaptFromTask<DigitizerSpec>(calibloader)},
+    AlgorithmSpec{o2::framework::adaptFromTask<DigitizerSpec>(calibloader, requireCTPInput)},
     Options{
       {"pileup", VariantType::Int, 1, {"whether to run in continuous time mode"}},
       {"debug-stream", VariantType::Bool, false, {"Enable debug streaming"}}}

--- a/Steer/DigitizerWorkflow/src/CTPDigitizerSpec.cxx
+++ b/Steer/DigitizerWorkflow/src/CTPDigitizerSpec.cxx
@@ -35,7 +35,7 @@ class CTPDPLDigitizerTask : public o2::base::BaseDPLDigitizer
   using GRP = o2::parameters::GRPObject;
 
  public:
-  CTPDPLDigitizerTask() : o2::base::BaseDPLDigitizer(), mDigitizer() {}
+  CTPDPLDigitizerTask(const std::vector<o2::detectors::DetID>& detList) : o2::base::BaseDPLDigitizer(), mDigitizer(), mDetList(detList) {}
   ~CTPDPLDigitizerTask() override = default;
   void initDigitizerTask(framework::InitContext& ic) override
   {
@@ -44,23 +44,36 @@ class CTPDPLDigitizerTask : public o2::base::BaseDPLDigitizer
   void run(framework::ProcessingContext& pc)
   {
     // read collision context from input
-    //auto context = pc.inputs().get<o2::steer::DigitizationContext*>("collisioncontext");
-    //const bool withQED = context->isQEDProvided();
-    //auto& timesview = context->getEventRecords(withQED);
+    // auto context = pc.inputs().get<o2::steer::DigitizationContext*>("collisioncontext");
+    // const bool withQED = context->isQEDProvided();
+    // auto& timesview = context->getEventRecords(withQED);
     // read ctp inputs from input
-    auto ft0inputs = pc.inputs().get<gsl::span<o2::ft0::DetTrigInput>>("ft0");
-    auto fv0inputs = pc.inputs().get<gsl::span<o2::fv0::DetTrigInput>>("fv0");
 
     std::vector<o2::ctp::CTPInputDigit> finputs;
     TStopwatch timer;
     timer.Start();
     LOG(info) << "CALLING CTP DIGITIZATION";
-    // Input order: T0, V0, ... but O need also poisition of inputs DETInputs
-    for (const auto& inp : ft0inputs) {
-      finputs.emplace_back(CTPInputDigit{inp.mIntRecord, inp.mInputs, o2::detectors::DetID::FT0});
+    // Input order: T0, V0, ... but O need also position of inputs DETInputs
+    // fv0
+    if (std::find(mDetList.begin(), mDetList.end(), o2::detectors::DetID::FT0) != mDetList.end()) {
+      auto ft0inputs = pc.inputs().get<gsl::span<o2::ft0::DetTrigInput>>("ft0");
+      for (const auto& inp : ft0inputs) {
+        finputs.emplace_back(CTPInputDigit{inp.mIntRecord, inp.mInputs, o2::detectors::DetID::FT0});
+      }
     }
-    for (const auto& inp : fv0inputs) {
-      finputs.emplace_back(CTPInputDigit{inp.mIntRecord, inp.mInputs, o2::detectors::DetID::FV0});
+    // fv0
+    if (std::find(mDetList.begin(), mDetList.end(), o2::detectors::DetID::FV0) != mDetList.end()) {
+      auto fv0inputs = pc.inputs().get<gsl::span<o2::fv0::DetTrigInput>>("fv0");
+      for (const auto& inp : fv0inputs) {
+        finputs.emplace_back(CTPInputDigit{inp.mIntRecord, inp.mInputs, o2::detectors::DetID::FV0});
+      }
+    }
+    // emc
+    if (std::find(mDetList.begin(), mDetList.end(), o2::detectors::DetID::EMC) != mDetList.end()) {
+      auto emcinputs = pc.inputs().get<gsl::span<o2::ctp::CTPInputDigit>>("emc");
+      for (const auto& inp : emcinputs) {
+        finputs.emplace_back(inp);
+      }
     }
     gsl::span<CTPInputDigit> ginputs(finputs);
     auto digits = mDigitizer.process(ginputs);
@@ -76,16 +89,21 @@ class CTPDPLDigitizerTask : public o2::base::BaseDPLDigitizer
  protected:
   o2::parameters::GRPObject::ROMode mROMode = o2::parameters::GRPObject::PRESENT;
   o2::ctp::Digitizer mDigitizer; ///< Digitizer
+  std::vector<o2::detectors::DetID> mDetList;
 };
 o2::framework::DataProcessorSpec getCTPDigitizerSpec(int channel, std::vector<o2::detectors::DetID>& detList, bool mctruth)
 {
   std::vector<InputSpec> inputs;
   std::vector<OutputSpec> output;
+
   if (std::find(detList.begin(), detList.end(), o2::detectors::DetID::FT0) != detList.end()) {
     inputs.emplace_back("ft0", "FT0", "TRIGGERINPUT", 0, Lifetime::Timeframe);
   }
   if (std::find(detList.begin(), detList.end(), o2::detectors::DetID::FV0) != detList.end()) {
     inputs.emplace_back("fv0", "FV0", "TRIGGERINPUT", 0, Lifetime::Timeframe);
+  }
+  if (std::find(detList.begin(), detList.end(), o2::detectors::DetID::EMC) != detList.end()) {
+    inputs.emplace_back("emc", "EMC", "TRIGGERINPUT", 0, Lifetime::Timeframe);
   }
   output.emplace_back("CTP", "DIGITS", 0, Lifetime::Timeframe);
   output.emplace_back("CTP", "ROMode", 0, Lifetime::Timeframe);
@@ -93,7 +111,7 @@ o2::framework::DataProcessorSpec getCTPDigitizerSpec(int channel, std::vector<o2
     "CTPDigitizer",
     inputs,
     output,
-    AlgorithmSpec{adaptFromTask<CTPDPLDigitizerTask>()},
+    AlgorithmSpec{adaptFromTask<CTPDPLDigitizerTask>(detList)},
     Options{{"pileup", VariantType::Int, 1, {"whether to run in continuous time mode"}},
             {"disable-qed", o2::framework::VariantType::Bool, false, {"disable QED handling"}}}};
 }

--- a/Steer/DigitizerWorkflow/src/SimpleDigitizerWorkflow.cxx
+++ b/Steer/DigitizerWorkflow/src/SimpleDigitizerWorkflow.cxx
@@ -197,6 +197,9 @@ void customize(std::vector<o2::framework::ConfigParamSpec>& workflowOptions)
   // option to use/not use CCDB for EMCAL
   workflowOptions.push_back(ConfigParamSpec{"no-use-ccdb-emc", o2::framework::VariantType::Bool, false, {"Disable access to ccdb EMCAL simulation objects"}});
 
+  // option to require/not require CTP MB inputs in EMCAL
+  workflowOptions.push_back(ConfigParamSpec{"no-require-ctpinputs-emc", o2::framework::VariantType::Bool, false, {"Disable requirement of CTP min. bias inputs in EMCAL simulation"}});
+
   // option to use or not use the Trap Simulator after digitisation (debate of digitization or reconstruction is for others)
   workflowOptions.push_back(ConfigParamSpec{"disable-trd-trapsim", VariantType::Bool, false, {"disable the trap simulation of the TRD"}});
   workflowOptions.push_back(ConfigParamSpec{"trd-digit-downscaling", VariantType::Int, 1, {"only keep TRD digits for every n-th trigger"}});
@@ -648,9 +651,10 @@ WorkflowSpec defineDataProcessing(ConfigContext const& configcontext)
   // the EMCal part
   if (isEnabled(o2::detectors::DetID::EMC)) {
     auto useCCDB = !configcontext.options().get<bool>("no-use-ccdb-emc");
+    bool requireCTPInputs = !configcontext.options().get<bool>("no-require-ctpinputs-emc");
     detList.emplace_back(o2::detectors::DetID::EMC);
     // connect the EMCal digitization
-    digitizerSpecs.emplace_back(o2::emcal::getEMCALDigitizerSpec(fanoutsize++, mctruth, useCCDB));
+    digitizerSpecs.emplace_back(o2::emcal::getEMCALDigitizerSpec(fanoutsize++, requireCTPInputs, mctruth, useCCDB));
     // connect the EMCal digit writer
     writerSpecs.emplace_back(o2::emcal::getEMCALDigitWriterSpec(mctruth));
   }

